### PR TITLE
Fix/nodes templates

### PR DESCRIPTION
**Pull request by qingyashu**
### Improvements
Allow user to download zip package including all selected nodes' tsv/json templates. Will omit `program` and `project` nodes' templates.

### CODE REVIEW

**Comment by qingyashu** [comment]
Branch deployed in https://qingya.planx-pla.net/DD, please have a try.

**Comment by zflamig** [comment]
Nitpicky UI request, can we add a little bit more space between the "Download Templates" and the ... on the button?

The files are numbered incorrectly (in reverse order) when downloading a node with multiple links too.

File: src/DataDictionary/DataModelStructure/DataModelStructure.jsx [hunk 1]
```diff
@@ -15,6 +17,43 @@ class DataModelStructure extends React.Component {
     this.props.onSetOverlayPropertyTableHidden(!this.props.overlayPropertyHidden);
   };
 
+  downloadTemplatesEnabled = () => {
+    const nodeIDs = [];
+    this.props.dataModelStructure.forEach((entry) => {
+      const { nodeID, nodeIDsBefore } = entry;
+      nodeIDs.push(nodeID);
+      nodeIDs.concat(nodeIDsBefore);
```
[the remark this answers is not in the document]
Comment (qingyashu) [reply src/DataDictionary/DataModelStructure/DataModelStructure.jsx]: haha yes I forget concat is pure, and I need to exclude `nodeIDsBefore`~ will change this.

File: src/DataDictionary/DataModelStructure/DataModelStructure.jsx [hunk 2]
```diff
@@ -15,6 +17,43 @@ class DataModelStructure extends React.Component {
...
+        nodeIDsBeforeForDownload.forEach((nid) => {
+          if (nid in nodesToDownload) return;
+          if (nodeIDsBeforeForDownload.length > 1) {
+            nodesToDownload[nid] = `${counter}-${innerCount}_${nid}_template.${format}`;
+            innerCount += 1;
+          } else nodesToDownload[nid] = `${counter}_${nid}_template.${format}`;
+        });
+        counter += 1;
+      }
+      if (!this.props.excludedNodesForTemplates.includes(nodeID)) {
```
[the remark this answers is not in the document]
Comment (qingyashu) [reply src/DataDictionary/DataModelStructure/DataModelStructure.jsx]: There are only `project` and `program` two nodes to exclude, so it won't take too many lookup time. I've also checked the 3 lookup places and seems they are used for different arrays so there's no way I can reduce them into common function  :(  But I can add more comments to make the logic more readable~

[the remark this answers is not in the document]
Comment (qingyashu) [reply src/DataDictionary/DataModelStructure/DataModelStructure.jsx]: or I'll make this line into a small function `this.props.excludedNodesForTemplates.includes(nodeID)` :p

**Comment by qingyashu** [comment]
Thanks for comments! Code updated and deployed in https://qingya.planx-pla.net/DD, please have a check again.

**Comment by ZakirG** [comment]
Looks great, works very well! Happy with the code comments.

File: src/DataDictionary/utils.js [hunk 3]
```diff
@@ -107,6 +113,11 @@ export const downloadMultiTemplate = (format, nodesToDownload) => {
       throw new Error(`cannot download template for node "${nodeID}"`);
     });
   })).then(() => {
+    const time = new Date();
+    const startingNodeName = 'project';
+    let readmeContent = `The following TSV templates were downloaded from ${appname} on ${time.toLocaleDateString()} ${time.toLocaleTimeString()}. The following are all possible paths from ${startingNodeName} node to ${clickingNodeName} using data dictionary version ${dictionaryVersion}. The downloaded ${format} files need to be submitted in the order shown in the chosen path(s) below.\n`;
```
Comment (mfshao) [inline src/DataDictionary/utils.js]: `The following TSV templates were downloaded` should be `These ${format} templates were downloaded`

Comment (mfshao) [reply src/DataDictionary/utils.js]: Also, I think we could either add a number for each path in the list below, or add something like `There are {$totalPathNumber} paths in total.` before ` The downloaded ${format} files need to be submitted in the order shown in the chosen path(s) below.` in here.

Comment (qingyashu) [reply src/DataDictionary/utils.js]: thanks! will add path number~

**Pull request merged**